### PR TITLE
Backport/buildtool server side nbt (#777)

### PR DESCRIPTION
**Pull request by xavierh**
Backport from version/1.11
* Add a persistent UUID to the server, saved in minecolomies.dat
Send the UUID to the client when the player log in.
The UUID will be use to create a directory on the client to store data specific to a server

* request as schematic from the client and send the schematic back by the server

* Allow to override huts and decorations by nbt file in a directory

* Add custom decoration (from local player)
Cache file using the md5 hash as a file name inside cache directory
When the player want to use custom, he will need to send the schematic unless the cache/<md5 hash> already exists

Schematic can be send to the server only if enabled in the configuration (disable by default), no schematic management is done yet, no limitation on the number of schamatic send, unused one are still there.

* Remember the last structure build, so we can quicky build it again, usefull for walls
Store the name instead of index as they may change

* Allow to rename/delete custom schematic

Review please

### CODE REVIEW

**Comment by xavierh** [comment]
Build but not tested yet

File: src/main/java/com/minecolonies/coremod/colony/Structures.java [hunk 1]
```diff
@@ -60,200 +113,822 @@ public static void init()
...
+        }
+        return new ArrayList<>();
     }
 
     /**
-     * Returns a list of styles for one specific decoration.
+     * Class to handle schematic naming.
+     * It does extract information from a schematic using its name.
+     */
+    public static class StructureName
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/Structures.java:483]: MAJOR Reduce this class from 118 to the maximum allowed 25 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/util/BlockUtils.java [hunk 2]
```diff
@@ -214,7 +215,11 @@ public static ItemStack getItemStackFromBlockState(@NotNull final IBlockState bl
 
     private static Item getItem(@NotNull final IBlockState blockState)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/BlockUtils.java:216]: MAJOR This method has 146 lines, which is greater than the 100 lines authorized. Split it into smaller methods. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS138)
MAJOR Reduce the number of returns of this method 33, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "getItem" is 42 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Refactor this method to reduce its Cognitive Complexity from 67 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 3]
```diff
@@ -84,61 +98,312 @@
...
+            return Structure.getStreamFromFolder(Structure.getClientSchematicsFolder(), structureName);
+        }
+        else if (!Structures.SCHEMATICS_PREFIX.equals(sn.getPrefix()))
+        {
+            return null;
+        }
+        else
+        {
+            //Look in the folder first
+            inputstream = Structure.getStreamFromFolder(MineColonies.proxy.getSchematicsFolder(), structureName);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:212]: BLOCKER Close this "InputStream". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 4]
```diff
@@ -84,61 +98,312 @@
...
+     * @return the input stream or null
+     */
+    @Nullable
+    public static InputStream getStream(final String structureName)
+    {
+        final Structures.StructureName sn = new Structures.StructureName(structureName);
+        InputStream inputstream = null;
+        if (Structures.SCHEMATICS_CACHE.equals(sn.getPrefix()))
+        {
+            return Structure.getStreamFromFolder(Structure.getCachedSchematicsFolder(), structureName);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:199]: BLOCKER Close this "InputStream". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 5]
```diff
@@ -84,61 +98,312 @@
...
-                    decorationFolder = new File(Minecraft.getMinecraft().mcDataDir, "minecolonies/");
-                }
-                else
-                {
-                    decorationFolder = new File(FMLCommonHandler.instance().getMinecraftServerInstance().getDataDirectory(), "minecolonies/");
-                }
-                if(decorationFolder.exists())
+        if (inputStream == null)
+        {
+            inputStream = Structure.getStream(correctStructureName);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:121]: BLOCKER Close this "InputStream". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 6]
```diff
@@ -603,22 +604,22 @@ private void initStructure(final Node mineNode, final int rotateTimes, BlockPos
     {
         if(mineNode == null)
         {
-            loadStructure("miner/minerMainShaft", getRotationFromVector(), structurePos, false);
+            loadStructure(Structures.SCHEMATICS_PREFIX + "/miner/minerMainShaft", getRotationFromVector(), structurePos, false);
         }
         else
         {
             //todo we can add other nodeTypes without a problem.
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java:611]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 150 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![CRITICAL][CRITICAL] 4 critical
* ![MAJOR][MAJOR] 91 major
* ![MINOR][MINOR] 46 minor
* ![INFO][INFO] 6 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractEntityAIStructure.java#L256](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L256): Reduce the number of conditional operators (6) used in the expression (maximum allowed 5). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![CRITICAL][CRITICAL] [AbstractEntityAIStructure.java#L302](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L302): Reduce the number of conditional operators (7) used in the expression (maximum allowed 5). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![CRITICAL][CRITICAL] [AbstractEntityAIStructure.java#L822](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java#L822): Reduce the number of conditional operators (7) used in the expression (maximum allowed 5). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![CRITICAL][CRITICAL] [BlockUtils.java#L262](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/util/BlockUtils.java#L262): Reduce the number of conditional operators (6) used in the expression (maximum allowed 5). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.128 lines, which is greater than 750 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 40 to the maximum authorized 20 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): class "Colony" has 76 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L46](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L46): Refactor this class so it has no more than 20 fields, rather than the 33 it currently has. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1820)
1. ![MAJOR][MAJOR] [Colony.java#L580](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L580): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L1066](https://github.com/Minecolonies/minecolonies/blob/4521e43d27a5acfa2beefe9d4fc4484936f7625f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1066): The Cyclomatic Complexity of this method "onWorldTick" is 17 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by marvin-bitterlich** [comment]
Should be fine @Raycoms

**Comment by marvin-bitterlich** [comment]
closing in favor of #879

**Pull request closed without merging**